quarkchain/cluster/tests/test_failed_transactions_gas_fee.py

```diff
@@ -0,0 +1,112 @@
+from quarkchain.utils import token_id_encode
+from quarkchain.genesis import GenesisManager
+from quarkchain.cluster.tests.test_native_token import create_default_shard_state
+
+
+class TestFailTransactionGasCalucation(unittest.TestCase):
+    def setUp(self):
+        super().setUp()
+        config = get_test_env().quark_chain_config
+        self.root_coinbase = config.ROOT.COINBASE_AMOUNT
+        self.shard_coinbase = next(iter(config.shards.values())).COINBASE_AMOUNT
+        # to make test verification easier, assume following tax rate
+        assert config.REWARD_TAX_RATE == 0.5
+        self.tax_rate = config.reward_tax_rate  # type: Fraction
+        self.GENESIS_TOKEN = config.GENESIS_TOKEN  # type: str
+        self.genesis_token = config.genesis_token  # type: int
+
+    def getAfterTaxReward(self, value: int) -> int:
+        return value * self.tax_rate.numerator // self.tax_rate.denominator
+
+    def test_failed_transaction_gas(self):
+        """in-shard revert contract transaction validating the failed transaction gas used
+        """
+        QKC = token_id_encode("QKC")
+        id1 = Identity.create_random_identity()
+        acc1 = Address.create_from_identity(id1, full_shard_key=0)
+        acc2 = Address.create_random_account(full_shard_key=0)
+
+        env = get_test_env(
+            genesis_account=acc1,
+            genesis_minor_token_balances={
+                self.GENESIS_TOKEN: 200 * 10 ** 18,
+            },
+        )
+
+        state = create_default_shard_state(env=env)
+
+        # 1. creat failed contract with revert operation
+
+        CONTRACT_CREATION_WITH_REVERT_BYTECODE = "6080604052348015600f57600080fd5b503373ffffffffffffffffffffffffffffffffffffffff167fa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa60405160405180910390a2600080fdfe"
+        """
+        pragma solidity ^0.5.1;
+        contract RevertContract {
+            event Hi(address indexed);
+            constructor() public {
+                emit Hi(msg.sender);
+                revert();
+            }
+            function f() public {
+                emit Hi(msg.sender);
+            }
+        }
+        """
+        # This transaction cost is calculated by remix, which is smaller than the opcodes.GTXCOST due to revert.
+        FAILED_TRANSACTION_COST = 60004
+    
+        tx = contract_creation_tx(
+            shard_state=state,
+            key=id1.get_key(),
+            from_address=acc1,
+            to_full_shard_key=acc1.full_shard_key,
+            bytecode=CONTRACT_CREATION_WITH_REVERT_BYTECODE,
+            gas_token_id=self.genesis_token,
+            transfer_token_id=self.genesis_token,
+        )
+        # Should succeed
+        self.assertTrue(state.add_tx(tx))
+        b1 = state.create_block_to_mine(address=acc2)
+        self.assertEqual(len(b1.tx_list), 1)
+
+        state.finalize_and_add_block(b1)
+
+        self.assertEqual(state.header_tip, b1.header)
```

Comment:
can also check receipts to make sure the tx failed. code example is here https://github.com/QuarkChain/pyquarkchain/blob/6a66a11ad48835163a20b3b06ed3ff59a35dde11/quarkchain/cluster/tests/test_shard_state.py#L240-L243